src/plasmapy/simulation/particle_tracker/particle_tracker.py

```diff
@@ -706,6 +799,55 @@ def _push(self) -> None:
 
         self.dt = dt
 
+        # Update velocities to reflect stopping
+        if not self._do_stopping:
+            return
+
+        # TODO: maybe break out different stopping models into separate functions?
+        #  particle_stopping_routines?
+        # S = np.zeros(self.nparticles_tracked) * u.J / u.m
+        #
+        # for grid in self.grids:
+        #     _S = grid.nearest_neighbor_interpolator(
+        #         pos_tracked * u.m, "S", persistent=False
+        #     )
+        #     _S = np.nan_to_num(_S, nan=0.0 * u.J / u.m)
+        #
+        #     S += _S
+        speeds = np.linalg.norm(self.v[tracked_mask], axis=-1)[:, np.newaxis]
+        dx = np.multiply(speeds, dt)
+        stopping_power = np.zeros((self.nparticles_tracked, 1))
+
+        for cs in self._stopping_power_interpolators:
+            # TODO: is this the proper way to handle the addition of multiple stopping powers?
+            relevant_kinetic_energy = (
+                self._particle_kinetic_energy[tracked_mask, np.newaxis] * u.J
+            )
+
+            interpolation_result = (
+                cs(relevant_kinetic_energy).to(u.J * u.m**2 / u.kg).value
+            )
+
+            stopping_power += interpolation_result
+
+        # Take the negative stopping power since we want to be removing energy
+        # from the particles
+
+        energy_loss_per_length = np.multiply(
+            stopping_power,
+            summed_field_values["rho"].to(u.kg / u.m**3).value[:, np.newaxis],
+        )
+        dE = -np.multiply(energy_loss_per_length, dx)
+
+        # Update the velocities of the particles using the new energy values
+        # TODO: again, figure out how to differentiate relativistic and classical cases
+        E = self._particle_kinetic_energy[tracked_mask, np.newaxis] + dE
+
+        speeds = np.sqrt(2 * E / self.m)
+        unit_vectors = np.multiply(1 / speeds, self.v[tracked_mask])
+
+        self.v[tracked_mask] = np.multiply(speeds, unit_vectors)
```

Comment:
I'm not sure I understand why the second mask is necessary? Particles that aren't tracked shouldn't be stopping, so they never need to be selected by this statement.

Reply:
The calculated change in energy (and the total energy by extension) is calculated only for the tracked particles (nparticles_tracked). We can use this to select out particles with negative energy but the mask we pass to stop_particles() needs to have shape nparticles. We construct this by initializing an empty book array of length nparticles and setting the indices within particles tracked equal to the E < 0 mask. 

Does that make more sense?

Reply:
Yes I think so - just a bit odd that stop_particles takes a length nparticles since the only the tracked ones can be stopped - unless it's also used for wire grids maybe?